Add chopsticks upgrade test to CI

timbrinded opened the pull request:
### What does it do?
Adds a new test action to the CI whenever we create a new runtime via the `Publish Runtime Draft` workflow

### What important points reviewers should know?
Uses chopsticks via Moonwall to test that current chains can be upgraded via existing runtimes. Very short execution time and will give quick feedback if storage migrations required etc.

### What alternative implementations were considered?
This is an additive step to the testing we already do on newly generated runtimes.

### What value does it bring to the blockchain users?
Increased network stability with guaranteed runtime upgrade testing.

crystalin commented on test/suites/rt-upgrade/test-upgrade-chain.ts:
```diff
@@ -0,0 +1,63 @@
+      timeout: 30000,
+      test: async function () {
+        const rtBefore = api.consts.system.version.specVersion.toNumber();
+        await context.upgradeRuntime(context);
```
This should be inside a before no ? cause otherwise if I run T3 only it won't test the runtime upgrade

timbrinded replied:
done

crystalin commented on test/suites/rt-upgrade/test-upgrade-chain.ts:
```diff
@@ -0,0 +1,63 @@
+      id: "T1",
+      title: "Calling chain constants data",
+      test: async () => {
+        const specVersion = api.consts.system.version.specVersion.toNumber();
```
Not sure what this tests

timbrinded replied:
Sanity check to make sure that:
1. Chopsticks is actually connected to *something*, and that it hasn't failed silently
2. Our polkadotJs provider is actually connected to a chain

I'll change this to look at the runtime name, as a less trite test.

crystalin replied:
I'd include that in the before and not being at test as itself, as it seems to be more a requirement than verifying a behavior

timbrinded replied:
Done

crystalin commented on test/suites/rt-upgrade/test-upgrade-chain.ts:
```diff
@@ -0,0 +1,63 @@
+        const rtBefore = api.consts.system.version.specVersion.toNumber();
+        await context.upgradeRuntime(context);
+        const rtafter = api.consts.system.version.specVersion.toNumber();
+        expect(rtBefore).toBeLessThan(rtafter);
```
This will fail for PR after the release is deployed but before the runtime bump version merge, as both the PR runtime version and the network version will be the same. I think it is acceptable (as it will be rare) but should probably be mentioned in the error message

timbrinded replied:
This is only getting run when `Publish Runtime Draft` workflow is run`, not on each pr, so this scenario is avoided entirely. 

I figured we would only want to run this test just before we are ready to cut a new RT version as tip of master will not always be prod-ready

crystalin replied:
Ok but if it is blocking the release, that might be an issue, as we sometime have to rush it.

crystalin replied:
Maybe having it run when merging to master a PR

crystalin commented on test/moonwall.config.json:
```diff
@@ -140,6 +140,72 @@
       ]
+    },
+    {
+      "name": "update_moonbeam",
```
Is there an easy way to run this locally ? As I don't think having the build/xxx.wasm file is standard, does moonwall allow to retrieve/copy it ?

timbrinded replied:
the naming scheme you have above is standard for when running through CI, when using the `download_artifact` github action.

For local runs you can always change moonwall config to `target/release/wbuild/moonbeam-runtime/moonbeam_runtime.compact.wasm` if you wished or `moonbeam-runtime-2201.wasm` as downloaded by moonwall.

timbrinded replied:
The CI does use the local generated artifacts like you wanted . I should clarify that when I mention download-arrifact, this is the way you share files between jobs (like how typescript tests use the binary from the build step).

timbrinded replied:
Done

crystalin reviewed (changes requested):
...

timbrinded commented:
Example of test run:
https://github.com/PureStake/moonbeam/actions/runs/4631356011/jobs/8194303222

:white_check_mark:  The test runs and works as expected
:x: The CI isn't being marked as a failure. 

I need to fix the moonwall reporter settings so that github actions can detect this failure.

timbrinded commented:
Ok this should be working now. here is the latest version which fails the CI (as expected) because current RT version is 2201 versus the most recent RT tag (also 2201).

https://github.com/PureStake/moonbeam/actions/runs/4632025252/jobs/8195791253#step:6:271

This should be good to merge now, and even in the case where the CI gives false negatives, it is non blocking so the RT release ticket will still be created automatically.

timbrinded commented:
@crystalin All the points you made have been addressed, we now will do a RT upgrade test everytime someone raises a PR. 

Whilst i was here I've also fixed and updated our existing CIs to remove Github warnings.

crystalin commented on .github/workflows/build.yml:
```diff
@@ -298,6 +298,26 @@ jobs:
           MB_VERSION=`./target/release/moonbeam --version`
           echo "Checking $MB_VERSION contains $GIT_COMMIT"
           echo "$MB_VERSION" | grep $GIT_COMMIT
+      - name: Save runtime wasm
```
Why change the order ?

timbrinded replied:
because running `cargo test` actually changes the RT that gets created in `/wbuild`!

The resulting blob is incompatible with uploading and the chain cannot be successfully upgraded using it.

timbrinded commented on .github/workflows/build.yml:
```diff
@@ -38,31 +38,31 @@ jobs:
           if [[ -n "${{ github.event.pull_request.head.sha }}" ]]; then
-            echo ::set-output name=git_branch::$(echo ${GITHUB_HEAD_REF})
-            echo ::set-output name=git_ref::${{ github.event.pull_request.head.sha }}
+            echo "git_branch=$(echo ${GITHUB_HEAD_REF})" >> $GITHUB_OUTPUT
```
Setoutput is being deprecated, the way I've done is how they suggest you do it in node beyond v12
